Теория:

9.Агрегирование и группировка:
https://docs.google.com/presentation/d/1rd8_dCD1rKBkYLUc29mm4VxHXYajV2uCUXWR7Nafjjk/edit#slide=id.ga89b2b8700_0_178

# Задание 6. Группировка футболистов

In [1]:
import pandas as pd

url = 'https://raw.githubusercontent.com/dm-fedorov/pandas_basic/master/data/football.csv'

df = pd.read_csv(url)
df.head() # groopby

Unnamed: 0               Name  Age Nationality                 Club  \
0           0           L. Messi   31   Argentina         FC Barcelona   
1           1  Cristiano Ronaldo   33    Portugal             Juventus   
2           2          Neymar Jr   26      Brazil  Paris Saint-Germain   
3           3             De Gea   27       Spain    Manchester United   
4           4       K. De Bruyne   27     Belgium      Manchester City   

       Value    Wage Position  Crossing  Finishing  ...  Penalties  Composure  \
0  110500000  565000       RF        84         95  ...         75         96   
1   77000000  405000       ST        84         94  ...         85         95   
2  118500000  290000       LW        79         87  ...         81         94   
3   72000000  260000       GK        17         13  ...         40         68   
4  102000000  355000      RCM        93         82  ...         79         88   

   Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  GKKicking  \
0       33              28             26         6          11         15   
1       28              31             23         7          11         15   
2       27              24             33         9           9         15   
3       15              21             13        90          85         87   
4       68              58             51        15          13          5   

   GKPositioning  GKReflexes  
0             14           8  
1             14          11  
2             15          11  
3             88          94  
4             10          13  

[5 rows x 42 columns]

Предположим, что перед нами стоит задача посчитать сумму зарплат по клубам, чтобы найти клуб с самой высокой зарплатой.

Мы можем сделать это с помощью функции `groupby`:

In [2]:
df.groupby('Club')

Мы передали на вход функции `groupby` колонку (или список колонок), по которой будет осуществляться группировка. Но в ответ получили что-то странное. Это объект группировки, он хранит в себе информацию о том, какие строки датафрейма (по индексным номерам) соответствуют определенной группе (в нашем случае определенному клубу). Увидеть это можно, вызвав у объекта группировки атрибут `groups`:

In [3]:
df.groupby('Club').groups

{' SSV Jahn Regensburg': [1642, 2284, 2374, 2401, 2575, 2583, 2850, 2868, 3418, 3908, 3989, 4154, 5151, 5177, 5525, 5885, 6070, 6679, 7173, 7477, 8905, 9061, 10825, 12482], '1. FC Heidenheim 1846': [1060, 1805, 1978, 2050, 2066, 2926, 3584, 3605, 4312, 4457, 4783, 6818, 7160, 7166, 7460, 9475, 10052, 10565, 11217, 12236], '1. FC Kaiserslautern': [2931, 3590, 4270, 4413, 4475, 4483, 4709, 4779, 5079, 5325, 5403, 5855, 5917, 7295, 8784, 9496, 9964, 10019, 10122, 10671, 11500, 11970, 12222], '1. FC Köln': [166, 379, 899, 1399, 3415, 6904, 7052, 8892, 12171, 12376], '1. FC Magdeburg': [1967, 2069, 2152, 2569, 2886, 3030, 3300, 3332, 4436, 4913, 5002, 5033, 5493, 5768, 6329, 7046, 7595, 8387, 9626, 11125, 11400], '1. FC Nürnberg': [724, 1208, 1254, 1920, 4775, 5265, 5556, 5985, 8590, 8632, 9217, 9858, 10480], '1. FC Union Berlin': [1422, 1493, 2530, 2876, 3635, 3844, 4683, 4826, 11038, 11296, 12340], '1. FSV Mainz 05': [545, 670, 765, 1201, 1342, 6218, 6868, 8792, 10293, 10561, 10566], 'AC 

Это означает, что `FC Köln` соответствуют следующие строки: `166, 379, 899, 1399, 3415, 6904, 7052, 8892, 12171, 12376`.

О чем нам это говорит? Нам — ни о чем, но для алгоритма группировки это важная информация. Он поймет, какие строки брать для включения в группу, когда мы решим применить к этим строкам какую-нибудь агрегирующую функцию, например функцию суммы `sum`:

In [4]:
grouped_df = df.groupby('Club').sum()
grouped_df.head()

Unnamed: 0  \
Club                                
 SSV Jahn Regensburg       129124   
1. FC Heidenheim 1846      117834   
1. FC Kaiserslautern       167351   
1. FC Köln                  54865   
1. FC Magdeburg            116731   

                                                                    Name  Age  \
Club                                                                            
 SSV Jahn Regensburg   P. PentkeS. NachreinerA. NandzikB. SallerA. Ge...  609   
1. FC Heidenheim 1846  M. SchnattererN. TheuerkaufM. WittekT. Beerman...  473   
1. FC Kaiserslautern   T. ThieleJ. LöhmannsröbenH. ZuckA. HainaultJ. ...  544   
1. FC Köln             T. HornJ. HectorL. SchaubM. LehmannM. BaderC. ...  225   
1. FC Magdeburg        C. BeckJ. FejzićR. WeilS. SchäferN. HammannR. ...  515   

                                                             Nationality  \
Club                                                                       
 SSV Jahn Regensburg   GermanyGermanyGermanyGermanyGermanyGermanyGerm...   
1. FC Heidenheim 1846  GermanyGermanyGermanyGermanyGermanyGermanyGerm...   
1. FC Kaiserslautern   GermanyGermanyGermanyCanadaGermanyGermanyGerma...   
1. FC Köln             GermanyGermanyAustriaGermanyGermanyGermanyDenm...   
1. FC Magdeburg        GermanyBosnia HerzegovinaGermanyGermanyGermany...   

                          Value   Wage  \
Club                                     
 SSV Jahn Regensburg   15195000  90000   
1. FC Heidenheim 1846  18290000  76000   
1. FC Kaiserslautern   11195000  33000   
1. FC Köln             46810000  92000   
1. FC Magdeburg        57525000  84000   

                                                                Position  \
Club                                                                       
 SSV Jahn Regensburg   GKCBLBRBRDMGKCDMCMLBRAMLMRCBRBLMCBSTLDMCBCBSTC...   
1. FC Heidenheim 1846        LMLBCBLCBRBGKCAMRBRCBRMSTRMSTCBRMCMLMCBGKLB   
1. FC Kaiserslautern   LSRDMLMLCBRSCDMRBRMRMLBCDMSTCBRBGKCBLWGKLMGKCD...   
1. FC Köln                                        GKLWBRMCDMRBCMCMCBGKGK   
1. FC Magdeburg           RSGKCBLCBCBLCMRMCBCBLSRWRCBGKRMRWLBRCMRMGKSTCF   

                       Crossing  Finishing  HeadingAccuracy  ...  Penalties  \
Club                                                         ...              
 SSV Jahn Regensburg       1104        944             1283  ...       1042   
1. FC Heidenheim 1846       961        860              955  ...        923   
1. FC Kaiserslautern       1106        918             1133  ...       1030   
1. FC Köln                  444        357              434  ...        431   
1. FC Magdeburg             948        903             1044  ...        938   

                       Composure  Marking  StandingTackle  SlidingTackle  \
Club                                                                       
 SSV Jahn Regensburg        1332     1205            1146           1088   
1. FC Heidenheim 1846       1093      888             952            890   
1. FC Kaiserslautern        1174      938             989            996   
1. FC Köln                   515      453             458            452   
1. FC Magdeburg             1057      933             930            892   

                       GKDiving  GKHandling  GKKicking  GKPositioning  \
Club                                                                    
 SSV Jahn Regensburg        405         406        368            409   
1. FC Heidenheim 1846       335         317        287            335   
1. FC Kaiserslautern        392         373        348            368   
1. FC Köln                  272         251        276            252   
1. FC Magdeburg             367         373        354            361   

                       GKReflexes  
Club                               
 SSV Jahn Regensburg          424  
1. FC Heidenheim 1846         306  
1. FC Kaiserslautern          356  
1. FC Köln                    268  
1. FC Magdeburg    

In [5]:
# лучше так:
grouped_df = df.groupby('Club').agg("sum")
grouped_df.head()

Unnamed: 0  \
Club                                
 SSV Jahn Regensburg       129124   
1. FC Heidenheim 1846      117834   
1. FC Kaiserslautern       167351   
1. FC Köln                  54865   
1. FC Magdeburg            116731   

                                                                    Name  Age  \
Club                                                                            
 SSV Jahn Regensburg   P. PentkeS. NachreinerA. NandzikB. SallerA. Ge...  609   
1. FC Heidenheim 1846  M. SchnattererN. TheuerkaufM. WittekT. Beerman...  473   
1. FC Kaiserslautern   T. ThieleJ. LöhmannsröbenH. ZuckA. HainaultJ. ...  544   
1. FC Köln             T. HornJ. HectorL. SchaubM. LehmannM. BaderC. ...  225   
1. FC Magdeburg        C. BeckJ. FejzićR. WeilS. SchäferN. HammannR. ...  515   

                                                             Nationality  \
Club                                                                       
 SSV Jahn Regensburg   GermanyGermanyGermanyGermanyGermanyGermanyGerm...   
1. FC Heidenheim 1846  GermanyGermanyGermanyGermanyGermanyGermanyGerm...   
1. FC Kaiserslautern   GermanyGermanyGermanyCanadaGermanyGermanyGerma...   
1. FC Köln             GermanyGermanyAustriaGermanyGermanyGermanyDenm...   
1. FC Magdeburg        GermanyBosnia HerzegovinaGermanyGermanyGermany...   

                          Value   Wage  \
Club                                     
 SSV Jahn Regensburg   15195000  90000   
1. FC Heidenheim 1846  18290000  76000   
1. FC Kaiserslautern   11195000  33000   
1. FC Köln             46810000  92000   
1. FC Magdeburg        57525000  84000   

                                                                Position  \
Club                                                                       
 SSV Jahn Regensburg   GKCBLBRBRDMGKCDMCMLBRAMLMRCBRBLMCBSTLDMCBCBSTC...   
1. FC Heidenheim 1846        LMLBCBLCBRBGKCAMRBRCBRMSTRMSTCBRMCMLMCBGKLB   
1. FC Kaiserslautern   LSRDMLMLCBRSCDMRBRMRMLBCDMSTCBRBGKCBLWGKLMGKCD...   
1. FC Köln                                        GKLWBRMCDMRBCMCMCBGKGK   
1. FC Magdeburg           RSGKCBLCBCBLCMRMCBCBLSRWRCBGKRMRWLBRCMRMGKSTCF   

                       Crossing  Finishing  HeadingAccuracy  ...  Penalties  \
Club                                                         ...              
 SSV Jahn Regensburg       1104        944             1283  ...       1042   
1. FC Heidenheim 1846       961        860              955  ...        923   
1. FC Kaiserslautern       1106        918             1133  ...       1030   
1. FC Köln                  444        357              434  ...        431   
1. FC Magdeburg             948        903             1044  ...        938   

                       Composure  Marking  StandingTackle  SlidingTackle  \
Club                                                                       
 SSV Jahn Regensburg        1332     1205            1146           1088   
1. FC Heidenheim 1846       1093      888             952            890   
1. FC Kaiserslautern        1174      938             989            996   
1. FC Köln                   515      453             458            452   
1. FC Magdeburg             1057      933             930            892   

                       GKDiving  GKHandling  GKKicking  GKPositioning  \
Club                                                                    
 SSV Jahn Regensburg        405         406        368            409   
1. FC Heidenheim 1846       335         317        287            335   
1. FC Kaiserslautern        392         373        348            368   
1. FC Köln                  272         251        276            252   
1. FC Magdeburg             367         373        354            361   

                       GKReflexes  
Club                               
 SSV Jahn Regensburg          424  
1. FC Heidenheim 1846         306  
1. FC Kaiserslautern          356  
1. FC Köln                    268  
1. FC Magdeburg    

Получился датафрейм, где все параметры, к которым можно было применить соответствующую агрегирующую функцию, были просуммированы. Если точнее, были просуммированы все численные параметры. Другие параметры, такие как национальность (`Nationality`), не будут отображены в датафрейме сгруппированных данных. Это связано с тем, что к таким колонкам нельзя применить агрегирующую функцию, потому что они содержат не численные параметры, а текст.

Давайте обратимся по индексу к какой-то из строк получившегося датафрейма:

In [6]:
grouped_df.loc['Ajax']

Unnamed: 0                                                     26041
Name               H. ZiyechM. de LigtF. de JongN. TagliaficoD. B...
Age                                                              275
Nationality        MoroccoNetherlandsNetherlandsArgentinaNetherla...
Value                                                      152250000
Wage                                                          158000
Position                               RAMRCBLDMLBLBRWSTCBRBCMLBLWCM
Crossing                                                         844
Finishing                                                        723
HeadingAccuracy                                                  848
ShortPassing                                                     940
Volleys                                                          696
Dribbling                                                        932
Curve                                                            779
FKAccuracy                        

Таким образом, мы получили серию, где индекс содержит все параметры, которые можно было просуммировать, а значения серии — суммы параметров для всех игроков `ФК "Ajax"`.

Теперь мы можем получить сумму зарплат игроков этого футбольного клуба:

In [7]:
grouped_df.loc['Ajax']['Wage']

158000

Тут важно заметить, что особого смысла суммировать другие параметры нет. Например, что нам даёт сумма возрастов? Это неинформативная метрика, а значит смысла её подсчитывать нет. Поэтому мы можем ограничить список наших параметров и при группировке осуществлять суммирование только по параметру `"Wage"` (заработная плата):

In [8]:
grouped_df = df.groupby('Club')['Wage'].sum()
grouped_df

Club
 SSV Jahn Regensburg      90000
1. FC Heidenheim 1846     76000
1. FC Kaiserslautern      33000
1. FC Köln                92000
1. FC Magdeburg           84000
                          ...  
Zagłębie Sosnowiec        27000
Çaykur Rizespor          118000
Örebro SK                 36000
Östersunds FK             39000
Śląsk Wrocław             50000
Name: Wage, Length: 650, dtype: int64

In [9]:
# лучше так:
grouped_df = df.groupby('Club').agg({'Wage': 'sum'})
grouped_df

Wage
Club                         
 SSV Jahn Regensburg    90000
1. FC Heidenheim 1846   76000
1. FC Kaiserslautern    33000
1. FC Köln              92000
1. FC Magdeburg         84000
...                       ...
Zagłębie Sosnowiec      27000
Çaykur Rizespor        118000
Örebro SK               36000
Östersunds FK           39000
Śląsk Wrocław           50000

[650 rows x 1 columns]

Уже лучше!

Но помните наш исходный вопрос: `"В каком клубе самая высокая зарплата?"`

Для ответа на этот вопрос нам достаточно лишь отсортировать полученные данные по убыванию суммы зарплат с помощью функции `sort_values`:

In [10]:
# параметр ascending=False указывает, что данные надо сортировать по убыванию
grouped_df = df \
               .groupby('Club') \
               .agg({'Wage': 'sum'}) \
               .sort_values(by="Wage", ascending=False)

grouped_df.head(5)

Wage
Club                      
Real Madrid        4138000
FC Barcelona       3967000
Manchester City    3097000
Manchester United  2357000
Juventus           2335000

И это `Real Madrid`, что вполне ожидаемо!

Давайте рассмотрим ещё несколько примеров использования функции `groupby`.

Например, построим такую таблицу, где сгруппируем игроков по национальностям (`Nationality`) и посчитаем среднюю зарплату, средний возраст и среднюю силу удара. Вдруг это всё как-то взаимосвязано?!

Для подсчета среднего мы будем использовать агрегирующую функцию `mean`:

In [11]:
df.groupby('Nationality')[['Wage','Age','ShotPower']].mean()

Wage        Age  ShotPower
Nationality                                    
Afghanistan   1250.000000  22.500000  53.750000
Albania       5931.034483  23.689655  54.551724
Algeria      15810.810811  27.027027  56.567568
Andorra       1000.000000  28.000000  58.000000
Angola        7100.000000  24.700000  63.700000
...                   ...        ...        ...
Uzbekistan    9000.000000  29.500000  71.500000
Venezuela     6611.111111  23.638889  57.222222
Wales         9529.411765  24.754902  48.558824
Zambia        2833.333333  21.833333  66.833333
Zimbabwe      5000.000000  26.555556  62.888889

[156 rows x 3 columns]

Далее отсортируем по убыванию усредненной зарплаты:

In [12]:
df \
  .groupby('Nationality')[['Wage','Age','ShotPower']] \
  .mean() \
  .sort_values('Wage', ascending=False) \
  .head(10)

Wage        Age  ShotPower
Nationality                                           
Dominican Republic  71000.000000  23.000000  75.500000
Egypt               35545.454545  25.818182  59.363636
Gabon               28900.000000  26.400000  56.900000
Croatia             26722.222222  24.819444  54.305556
Equatorial Guinea   25666.666667  28.000000  55.333333
Belgium             20024.390244  24.030488  56.390244
Ecuador             18333.333333  24.619048  60.666667
Uruguay             17590.361446  26.771084  56.192771
Brazil              17371.158392  27.898345  58.203310
Algeria             15810.810811  27.027027  56.567568

Что?! Доминиканская республика? Очень странно, давайте поближе посмотрим на данные по этой стране:

In [13]:
df.query("Nationality == 'Dominican Republic'")[['Name','Club','Wage','Age','ShotPower']]

Name         Club    Wage  Age  ShotPower
291      Mariano  Real Madrid  140000   24         85
6279  L. Quezada   Córdoba CF    2000   22         66

Теперь всё ясно. В нашей выборке только два футболиста из этой страны, один из них играет в клубе `"Real Madrid"` и получает 140000 евро, а другой в испанском клубе второго дивизиона с зарплатой в 70 раз ниже! Поэтому среднее и получилось таким высоким.

### Задача 1

Посчитайте среднюю зарплату (`Wage`) и цену (`Value`) игроков разных позиций (`Position`). Представители какой позиции имеют самую высокую среднюю цену?

In [14]:
average_values = df.groupby('Position').agg({'Wage': 'mean', 'Value': 'mean'})
position_with_highest_value = average_values['Value'].idxmax()

print("Позиция с самой высокой средней ценой игроков:", position_with_highest_value)

Позиция с самой высокой средней ценой игроков: LF


Какова средняя зарплата футболистов на данной позиции?

In [15]:
average_wage_highest_value_position = average_values.loc[position_with_highest_value, 'Wage']
print("Средняя зарплата футболистов на позиции с самой высокой средней ценой:", average_wage_highest_value_position)

Средняя зарплата футболистов на позиции с самой высокой средней ценой: 75125.0


В предыдущих примерах мы использовали такие агрегирующие функции, как `sum` (сумма) и `mean` (среднее). Давайте рассмотрим, какие ещё функции можно использовать при группировке данных.

Функция `nunique`, которая позволяет посчитать количество уникальных значений по серии. Её лучше всего применять к тем колонкам датафрейма, в которых хранятся категорийные данные:

In [16]:
df.groupby('Nationality')[['Club','Name']].nunique()

Club  Name
Nationality            
Afghanistan     4     4
Albania        26    29
Algeria        33    37
Andorra         1     1
Angola          9    10
...           ...   ...
Uzbekistan      2     2
Venezuela      34    36
Wales          56   101
Zambia          5     6
Zimbabwe        8     9

[156 rows x 2 columns]

Функция `count`, с помощью которой можно посчитать количество элементов в группе. То есть результат её будет такой же, как у функции `value_counts`:

In [17]:
df.groupby('Club')['Name'].count()

Club
 SSV Jahn Regensburg     24
1. FC Heidenheim 1846    20
1. FC Kaiserslautern     23
1. FC Köln               10
1. FC Magdeburg          21
                         ..
Zagłębie Sosnowiec       25
Çaykur Rizespor          22
Örebro SK                26
Östersunds FK            20
Śląsk Wrocław            24
Name: Name, Length: 650, dtype: int64

In [18]:
# либо так:
df \
  .groupby('Club') \
  .agg({'Name' : "count"})

Name
Club                       
 SSV Jahn Regensburg     24
1. FC Heidenheim 1846    20
1. FC Kaiserslautern     23
1. FC Köln               10
1. FC Magdeburg          21
...                     ...
Zagłębie Sosnowiec       25
Çaykur Rizespor          22
Örebro SK                26
Östersunds FK            20
Śląsk Wrocław            24

[650 rows x 1 columns]

Функция `median`, с помощью которой можно посчитать медианное значение:

In [19]:
df.groupby('Club')['Dribbling'].median()

Club
 SSV Jahn Regensburg     57.5
1. FC Heidenheim 1846    59.5
1. FC Kaiserslautern     54.0
1. FC Köln               60.5
1. FC Magdeburg          53.0
                         ... 
Zagłębie Sosnowiec       58.0
Çaykur Rizespor          62.0
Örebro SK                55.0
Östersunds FK            61.5
Śląsk Wrocław            54.5
Name: Dribbling, Length: 650, dtype: float64

In [20]:
# либо так:
df \
  .groupby('Club') \
  .agg({'Dribbling' : "median"})

Dribbling
Club                            
 SSV Jahn Regensburg        57.5
1. FC Heidenheim 1846       59.5
1. FC Kaiserslautern        54.0
1. FC Köln                  60.5
1. FC Magdeburg             53.0
...                          ...
Zagłębie Sosnowiec          58.0
Çaykur Rizespor             62.0
Örebro SK                   55.0
Östersunds FK               61.5
Śląsk Wrocław               54.5

[650 rows x 1 columns]

Функция `max`, с помощью которой можно посчитать максимальное значение внутри группы:

In [21]:
df.groupby('Club')['Strength'].max()

Club
 SSV Jahn Regensburg     92
1. FC Heidenheim 1846    86
1. FC Kaiserslautern     85
1. FC Köln               76
1. FC Magdeburg          86
                         ..
Zagłębie Sosnowiec       84
Çaykur Rizespor          81
Örebro SK                85
Östersunds FK            88
Śląsk Wrocław            85
Name: Strength, Length: 650, dtype: int64

In [22]:
# либо так:
df \
  .groupby('Club', as_index=False) \
  .agg({'Strength': 'max'})

Club  Strength
0      SSV Jahn Regensburg        92
1    1. FC Heidenheim 1846        86
2     1. FC Kaiserslautern        85
3               1. FC Köln        76
4          1. FC Magdeburg        86
..                     ...       ...
645     Zagłębie Sosnowiec        84
646        Çaykur Rizespor        81
647              Örebro SK        85
648          Östersunds FK        88
649          Śląsk Wrocław        85

[650 rows x 2 columns]

Функция `min`, с помощью которой можно посчитать минимальное значение внутри группы:

In [23]:
df.groupby('Club')['Balance'].min()

Club
 SSV Jahn Regensburg     29
1. FC Heidenheim 1846    22
1. FC Kaiserslautern     29
1. FC Köln               31
1. FC Magdeburg          19
                         ..
Zagłębie Sosnowiec       28
Çaykur Rizespor          39
Örebro SK                34
Östersunds FK            32
Śląsk Wrocław            31
Name: Balance, Length: 650, dtype: int64

In [24]:
# либо так:
df \
  .groupby('Club') \
  .agg({'Balance': "min"})

Balance
Club                          
 SSV Jahn Regensburg        29
1. FC Heidenheim 1846       22
1. FC Kaiserslautern        29
1. FC Köln                  31
1. FC Magdeburg             19
...                        ...
Zagłębie Sosnowiec          28
Çaykur Rizespor             39
Örebro SK                   34
Östersunds FK               32
Śląsk Wrocław               31

[650 rows x 1 columns]

### Задача 2

Посчитайте среднюю (`mean`) и медианную (`median`) зарплату (`Wage`) футболистов из разных клубов (`Club`). В скольких клубах средняя и медианная зарплаты совпадают?

Подсказка: чтобы в процессе группировки применить к данным одновременно две агрегирующие функции, необходимо указать их как аргументы метода `agg`:

```Python
df.groupby(столбец_для_группировки)[столбцы_для_отображения].agg(['функция_1', 'функция_2'])
```

либо:

```Python
df.groupby(столбец_для_группировки).agg({'столбец_для_отображения':['функция_1', 'функция_2'])
```

In [25]:
# Группируем данные по клубам и считаем среднюю и медианную зарплату
club_wage_stats = df.groupby('Club')['Wage'].agg(['mean', 'median'])

# Находим клубы, в которых средняя и медианная зарплаты совпадают
clubs_same_mean_median = club_wage_stats[club_wage_stats['mean'] == club_wage_stats['median']]

print("Клубы, в которых средняя и медианная зарплаты совпадают:")
print(clubs_same_mean_median)
print("Всего таких клубов:", len(clubs_same_mean_median))

Клубы, в которых средняя и медианная зарплаты совпадают:
                           mean   median
Club                                    
1. FC Magdeburg          4000.0   4000.0
AEK Athens               1000.0   1000.0
AJ Auxerre               2000.0   2000.0
Benevento                2000.0   2000.0
Bohemian FC              1000.0   1000.0
Boyacá Chicó FC          1000.0   1000.0
Bray Wanderers           1000.0   1000.0
Brøndby IF               6000.0   6000.0
CD Feirense              4000.0   4000.0
CD Tondela               2000.0   2000.0
Cork City                1000.0   1000.0
Cruzeiro                13000.0  13000.0
Deportivo Pasto          1000.0   1000.0
Dinamo Zagreb            1000.0   1000.0
Dundalk                  1000.0   1000.0
Dynamo Kyiv              1000.0   1000.0
Envigado FC              1000.0   1000.0
FC Energie Cottbus       1000.0   1000.0
FC Metz                  1500.0   1500.0
Grimsby Town             2000.0   2000.0
Göztepe SK               7000.0   7000.0


### Задача 3

Продолжаем работать с клубами, в которых средняя зарплата совпадает с медианной. Каков максимальный размер средней зарплаты в этой группе клубов?

In [26]:
# Группируем данные по клубам и считаем среднюю и медианную зарплату
club_wage_stats = df.groupby('Club')['Wage'].agg(['mean', 'median'])

# Находим клубы, в которых средняя и медианная зарплаты совпадают
clubs_same_mean_median = club_wage_stats[club_wage_stats['mean'] == club_wage_stats['median']]

# Находим максимальный размер средней зарплаты в этой группе клубов
max_avg_wage = clubs_same_mean_median['mean'].max()

print("Максимальный размер средней зарплаты в группе клубов, где средняя и медианная зарплаты совпадают:", max_avg_wage)

Максимальный размер средней зарплаты в группе клубов, где средняя и медианная зарплаты совпадают: 13000.0


Как называется клуб, где игроки получают такую зарплату?

In [27]:
# Находим название клуба с максимальной средней зарплатой
club_max_avg_wage = clubs_same_mean_median[clubs_same_mean_median['mean'] == max_avg_wage].index[0]

print("Название клуба, в котором игроки получают максимальный размер средней зарплаты в группе:", club_max_avg_wage)


Название клуба, в котором игроки получают максимальный размер средней зарплаты в группе: Cruzeiro


### Задача 4

С помощью функции groupby посчитайте сумму зарплат (`Wage`) футболистов клуба (`Club`) `"Chelsea"`.

In [28]:
# Группируем по клубу "Chelsea" и считаем сумму зарплат
chelsea_wages_sum = df[df['Club'] == 'Chelsea']['Wage'].sum()

print(f"Сумма зарплат футболистов клуба Chelsea: {chelsea_wages_sum}")

Сумма зарплат футболистов клуба Chelsea: 2035000


### Задача 5

Определите максимальную зарплату футболиста национальности (`Nationality`) Аргентина (`"Argentina"`) в возрасте `20 лет`.

In [29]:
# Фильтрация данных по условию
max_salary_argentina_20 = df[(df['Nationality'] == 'Argentina') & (df['Age'] == 20)]['Wage'].max()

print(f"Максимальная зарплата футболиста национальности Аргентина в возрасте 20 лет: {max_salary_argentina_20}")

Максимальная зарплата футболиста национальности Аргентина в возрасте 20 лет: 54000


### Задача 6

Определите максимальную зарплату футболиста национальности (`Nationality`) Аргентина (`"Argentina"`) в возрасте `30 лет`.

In [30]:
max_salary_argentina_30 = df[(df['Nationality'] == 'Argentina') & (df['Age'] == 30)]['Wage'].max()

print(f"Максимальная зарплата футболиста национальности Аргентина в возрасте 30 лет: {max_salary_argentina_30}")

Максимальная зарплата футболиста национальности Аргентина в возрасте 30 лет: 300000


### Задача 7

Определите минимальную зарплату футболиста национальности (`Nationality`) Аргентина (`"Argentina"`) в возрасте `30 лет`.

In [31]:
min_salary_argentina_30 = df[(df['Nationality'] == 'Argentina') & (df['Age'] == 30)]['Wage'].min()

print(f"Минимальная зарплата футболиста национальности Аргентина в возрасте 30 лет: {min_salary_argentina_30}")


Минимальная зарплата футболиста национальности Аргентина в возрасте 30 лет: 1000


### Задача 8

Определите максимальную силу (`Strength`) и баланс (`Balance`) среди игроков клуба (`Club`) `"FC Barcelona"` из Аргентины (`"Argentina"`). Ответ через точку с запятой без пробела.

In [32]:
max_strength_balance = df[(df['Club'] == 'FC Barcelona') & (df['Nationality'] == 'Argentina')][['Strength', 'Balance']].max(axis=0).to_string(index=False, header=False, na_rep=';')

print(max_strength_balance)


59
95
